Gltrim skip while looping

gerddie opened the pull request:
This is a series to alleviate the problems reported in https://github.com/apitrace/apitrace/issues/839

With these patches a new call flag is added to tag calls that may be dropped after the first time when the last frame is looping. gltrim will tag these calls, and the a new retrace flag is added `--loop-skip-tagged-calls` to actually skip these calls during replay with the final loop. 

Testing with the trace posted in #839 showed a *significant* impact on performance, so using this  flag in a context where the performance is analyzed may not be a good option. 

The nice thing about this approach is that one can simply re-trim already trimmed traces. 

An alternative would be to either inject  object destroy calls for objects that are created but not deleted in the last frame (This is probably not feasible at all in the retracer, because it would mean that it would have to deal with the different backends)  or to let gltrim scan the whole original trace to add these destroy calls to the last frame.

jrfonseca reviewed (commented):
I don't oppose this direction, but there are a few different approaches worth considering here.

First of all, I'd like to understand one thing: was the `blender-base.trace` referred by https://github.com/apitrace/apitrace/issues/839 created by gltrim?  Does this mean that gltrim is lumping initialization calls like shader/texture creation into the last frame?  It would be better it didn't, as there are so many ways trying to distill the initialization call from rendering calls can go wrong.   That is, it would be better if gltrim collapsed all initialization into a _frame 0_.  This would indirectly eliminate the need for `CALL_FLAG_IGNORE_IN_LOOP`.

If `CALL_FLAG_IGNORE_IN_LOOP` is still necessary, it's also possible to retrospectively add it  to all traces when parsing, either in `lib/trace/trace_parser_flags.cpp`, or even better, in `LastFrameLoopParser`.  I feel it's superior, since it means we can tweak the calls that get the flag without changing the traces (therefore, making git repos grow even more quickly.)

gerddie commented on frametrim/ft_dependecyobject.cpp:
```diff
@@ -473,6 +475,45 @@ DependecyObjectMap::getById(unsigned id) const
 
+static bool isNonRepeatCall(const std::string& name)
+{
+    const std::unordered_set<std::string> nonRepeateCalls = {
```
Fixed.

gerddie commented:
I've changed the code to be handled completely in gltrim, that is, like you suggested, the relevant calls are moved before the last frame so that no objects are created in this last frame that are not also deleted. Since I moved out shader compilation the impact on performance is still there, but I had to remove some buffer and texture setup calls, because otherwise one would also have to track the bind calls and duplicated them.

jrfonseca commented:
> I've changed the code to be handled completely in gltrim, that is, like you suggested, the relevant calls are moved before the last frame so that no objects are created in this last frame that are not also deleted.

Thanks.  Looks good AFAICT.

Sorry for the delay -- I read this on my mobile, bug failed to mark to follow up, and then slipped my mind.
